doc/cueconvert.1
.TH "cueconvert" "1"
.SH NAME
cueconvert \- convert files between CUE and TOC formats
.SH SYNOPSIS
.B cueconvert
[
.B \-i
.I format
|
.BR \-\-input\-format =\fIformat\fP
] [
.B \-o
.I format
|
.BR \-\-output\-format =\fIformat\fP
] [
.I infile
[
.I outfile
] ]
.br
.B cueconvert \-h | \-\-help
.br
.B cueconvert \-V | \-\-version
.SH DESCRIPTION
.B cueconvert
coverts files between the CUE and TOC formats, each of which are commonly
used by compact disc ripping software to denote track breakpoints and other
data.
.PP
The operands are optional; if
.I infile
is not specified,
.B cueconvert
reads from standard input, and if
.I outfile
is not specified, it writes to standard output.
.PP
If no filenames are specified, standard input is used, and an input format
option
.I must
be specified.
The complementary format is used for output.
.PP
If filenames are provided, but format options are not specified,
the input and output formats will be guessed based on each filename's
suffix (e.g.,
.I .cue
or
.IR .toc ).
This heuristic is case-insensitive.
.SH OPTIONS
.TP
.BR \-h ", " \-\-help
displays a usage message and exits.
.TP
.BR \-i " \fIformat\fP, " \-\-input\-format=\fIformat\fP
sets the expected format of the input file to
.IR format .
.TP
.BR \-o " \fIformat\fP, " \-\-output\-format=\fIformat\fP
sets the format of the generated output file to
.IR format .
.TP
.B \-V ", " \-\-version
displays version information and exits.
.PP
The option argument
.I format
must be either
.B cue
or
.BR toc .
.SH "EXIT STATUS"
.B cueconvert
exits with status zero if it successfully coverts the input file, and
nonzero if there were problems.
.SH AUTHOR
Cuetools was written by Svend Sorensen.
Branden Robinson contributed fixes and enhancements to the utilities and
documentation.
.SH "SEE ALSO"
.BR cuebreakpoints(1),
.BR cueprint(1)


NEWS
Changes since 1.3

* Applied most of Branden Robinson's Debian patch
  (cuetools_1.3-2.diff.gz). This included many fixes and enhancements
  to tools and manual pages. cuebreakpoints and cueconvert will exit if
  they fail on any of the input files. cuebreakpoints, cueconvert, and
  cuetools will exit with a nonzero return value if there was a failure
  on an input file.

* Added version option to all tools.

* Single quoted string support has been removed.  This was broken, and I am not
  sure of any software that supports single quoted strings in cue or toc files.

Changes since 1.2

* Build system has been rewritten using automake/autoconf.

* Added append, prepend, and split pregap modes to cuebreakpoints.

* Fixed possible compilation error with Bison.

* All tools accept long options.

* Programs exit if --input-format or --output-format is an illegal value.

Changes since 1.1

* cuebreakpoints was not printing the last track breakpoint.  This has been
  fixed.

* cueprint now prints a zero-length string for undefined values.  Flags, width,
  and precision are applied to conversion characters which have no conversion.

* Added a tagging script, cuetag.sh, which uses cueprint, vorbiscomment, and
  mp3info to tag Ogg Vorbis and MP3 files.  This needs more testing and
  documentation.

* Added a track selection flag to cueprint.

Changes since 1.0

* File formats documentation has been readded and updated (docs/formats.txt).

* ``CDTEXTFILE`` is now a recognized (but ignored) keyword.

* If only a disc or track template is specified on the command line, then only
  the disc or track information is printed, respectively.

* cueprint now supports flags and width and precision values for template
  expansion (like the printf command).

Changes since 0.6

* cuetools-1.x is a significant code rewrite of 0.x.  The cue/toc parsing code
  has been reimplemented in lex/yacc.

* cuerename has been replaced by the more general utility cueprint.

* Usage syntax of the utilities has changed.  See the respective man pages.


src/tools/cueconvert.c
/*
 * cueconvert.c -- convert between cue/toc formats
 *
 * Copyright (C) 2004, 2005, 2006, 2007 Svend Sorensen
 * For license terms, see the file COPYING in this distribution.
 */

#include <getopt.h>	/* getopt_long() */
#include <stdio.h>	/* fprintf(), printf(), snprintf(), stderr */
#include <stdlib.h>	/* exit() */
#include <string.h>	/* strcasecmp() */
#include "cuefile.h"

#if HAVE_CONFIG_H
#include "config.h"
#else /* not HAVE_CONFIG_H */
#define PACKAGE_STRING "cueconvert"
#endif /* HAVE_CONFIG_H */

char *progname;

/* Print usage information and exit */
void usage(int status)
{
	if (0 == status) {
		printf("Usage: %s [option...] [infile [outfile]]\n", progname);
		printf("Convert file between the CUE and TOC formats.\n"
		       "\n"
		       "OPTIONS\n"
		       "-h, --help 			print usage\n"
		       "-i, --input-format cue|toc	set format of input file\n"
		       "-o, --output-format cue|toc	set format of output file\n"
		       "-V, --version			print version information\n");
	} else {
		fprintf(stderr, "Try `%s --help' for more information.\n", progname);
	}

	exit (status);
}

/* Print version information and exit */
void version()
{
	printf("%s\n", PACKAGE_STRING);

	exit(0);
}

int convert(char *iname, int iformat, char *oname, int oformat)
{
	Cd *cd = NULL;

	if (NULL == (cd = cf_parse(iname, &iformat))) {
		fprintf(stderr, "%s: error: unable to parse input file"
		        " `%s'\n", progname, iname);
		return -1;
	}

	if (UNKNOWN == oformat) {
		/* first use file suffix */
		if (UNKNOWN == (oformat = cf_format_from_suffix(oname))) {
			/* then use opposite of input format */
			switch(iformat) {
			case CUE:
					oformat = TOC;
					break;
			case TOC:
					oformat = CUE;
					break;
			}
		}
	}

	return cf_print(oname, &oformat, cd);
}

int main(int argc, char *argv[])
{
	int iformat = UNKNOWN;
	int oformat = UNKNOWN;
	int ret = 0;		/* return value of convert() */

	/* option variables */
	int c;
	/* getopt_long() variables */
	extern char *optarg;
	extern int optind;

	static struct option longopts[] = {
		{"help", no_argument, NULL, 'h'},
		{"input-format", required_argument, NULL, 'i'},
		{"output-format", required_argument, NULL, 'o'},
		{"version", no_argument, NULL, 'V'},
		{NULL, 0, NULL, 0}
	};

	progname = argv[0];

	while (-1 != (c = getopt_long(argc, argv, "hi:o:V", longopts, NULL))) {
		switch (c) {
		case 'h':
			usage(0);
			break;
		case 'i':
			if (0 == strcmp("cue", optarg)) {
				iformat = CUE;
			} else if (0 == strcmp("toc", optarg)) {
				iformat = TOC;
			} else {
				fprintf(stderr, "%s: error: unknown input file"
				        " format `%s'\n", progname, optarg);
				usage(1);
			}
			break;
		case 'o':
			if (0 == strcmp("cue", optarg)) {
				oformat = CUE;
			} else if (0 == strcmp("toc", optarg)) {
				oformat = TOC;
			} else {
				fprintf(stderr, "%s: error: unknown output file"
				        " format `%s'\n", progname, optarg);
				usage(1);
			}
			break;
		case 'V':
			version();
			break;
		default:
			usage(1);
			break;
		}
	}

	/* What we do depends on the number of operands. */
	if (optind == argc) {
		/* No operands: report breakpoints of stdin. */
		ret = convert("-", iformat, "-", oformat);
	} else if (optind == argc - 1) {
		/* One operand: convert operand file to stdout. */
		ret = convert(argv[optind], iformat, "-", oformat);
	} else if (optind == argc - 2) {
		/* Two operands: convert input file to output file. */
		ret = convert(argv[optind], iformat, argv[optind + 1], oformat);
	} else {
		usage(1);
	}

	return ret;
}
